diff --git a/bin/asdf b/bin/asdf
index a0962d8..cba5cf5 100755
--- a/bin/asdf
+++ b/bin/asdf
@@ -29,7 +29,8 @@ find_cmd() {
 
 find_asdf_cmd() {
   local asdf_cmd_dir
-  asdf_cmd_dir="$(asdf_dir)/lib/commands"
+  get_asdf_dir
+  asdf_cmd_dir="$REPLY/lib/commands"
   case "$1" in
   'exec' | 'current' | 'env' | 'global' | 'install' | 'latest' | 'local' | \
     'reshim' | 'uninstall' | 'update' | 'where' | 'which' | \
@@ -81,7 +82,8 @@ asdf_cmd() {
   ASDF_DATA_DIR=$(asdf_data_dir)
   export ASDF_DATA_DIR
 
-  ASDF_DIR=$(asdf_dir)
+  get_asdf_dir
+  ASDF_DIR=$REPLY
   export ASDF_DIR
 
   local result=
@@ -114,7 +116,8 @@ asdf_cmd() {
     . "$ASDF_CMD_FILE"
   else
     local asdf_cmd_dir
-    asdf_cmd_dir="$(asdf_dir)/lib/commands"
+    get_asdf_dir
+    asdf_cmd_dir="$REPLY/lib/commands"
     printf "%s\n" "Unknown command: \`asdf ${*}\`" >&2
     # shellcheck source=lib/commands/command-help.bash
     . "$asdf_cmd_dir/command-help.bash" >&2
diff --git a/lib/commands/command-help.bash b/lib/commands/command-help.bash
index 5453185..22bbaec 100644
--- a/lib/commands/command-help.bash
+++ b/lib/commands/command-help.bash
@@ -4,7 +4,8 @@
 
 asdf_help() {
   printf "version: %s\n\n" "$(asdf_version)"
-  cat "$(asdf_dir)/help.txt"
+  get_asdf_dir
+  cat "$REPLY/help.txt"
 }
 
 asdf_moto() {
diff --git a/lib/commands/command-plugin-test.bash b/lib/commands/command-plugin-test.bash
index 791ff5f..e825905 100644
--- a/lib/commands/command-plugin-test.bash
+++ b/lib/commands/command-plugin-test.bash
@@ -68,9 +68,11 @@ plugin_test_command() {
   fi
 
   TEST_DIR=$(mktemp -dt asdf.XXXX)
-  cp -R "$(asdf_dir)/bin" "$TEST_DIR"
-  cp -R "$(asdf_dir)/lib" "$TEST_DIR"
-  cp "$(asdf_dir)/asdf.sh" "$TEST_DIR"
+  get_asdf_dir
+  local asdf_dir=$REPLY
+  cp -R "$asdf_dir/bin" "$TEST_DIR"
+  cp -R "$asdf_dir/lib" "$TEST_DIR"
+  cp "$asdf_dir/asdf.sh" "$TEST_DIR"
 
   plugin_test() {
     export ASDF_DIR=$TEST_DIR
diff --git a/lib/commands/command-update.bash b/lib/commands/command-update.bash
index f991264..eb1e20b 100644
--- a/lib/commands/command-update.bash
+++ b/lib/commands/command-update.bash
@@ -4,7 +4,9 @@ update_command() {
   local update_to_head=$1
 
   (
-    cd "$(asdf_dir)" || exit 1
+    get_asdf_dir
+    local asdf_dir=$REPLY
+    cd "$asdf_dir" || exit 1
 
     if [ -f asdf_updates_disabled ] || ! git rev-parse --is-inside-work-tree &>/dev/null; then
       printf "Update command disabled. Please use the package manager that you used to install asdf to upgrade asdf.\n"
diff --git a/lib/commands/reshim.bash b/lib/commands/reshim.bash
index 82e6614..173817d 100644
--- a/lib/commands/reshim.bash
+++ b/lib/commands/reshim.bash
@@ -104,10 +104,12 @@ EOF
     grep '^#\sasdf-plugin:\s' <"$shim_path" >>"$temp_versions_path"
   fi
 
+  get_asdf_dir
+  local asdf_dir=$REPLY
   cat <<EOF >"$shim_path"
 #!/usr/bin/env bash
 $(sort -u <"$temp_versions_path")
-exec $(asdf_dir)/bin/asdf exec "${executable_name}" "\$@" # asdf_allow: ' asdf '
+exec $REPLY/bin/asdf exec "${executable_name}" "\$@" # asdf_allow: ' asdf '
 EOF
 
   rm "$temp_versions_path"
diff --git a/lib/utils.bash b/lib/utils.bash
index 5979974..a876212 100644
--- a/lib/utils.bash
+++ b/lib/utils.bash
@@ -7,9 +7,11 @@ GREP_COLORS=
 
 asdf_version() {
   local version git_rev
-  version="v$(cat "$(asdf_dir)/version.txt")"
-  if [ -d "$(asdf_dir)/.git" ]; then
-    git_rev="$(git --git-dir "$(asdf_dir)/.git" rev-parse --short HEAD)"
+  get_asdf_dir
+  local asdf_dir=$REPLY
+  version="v$(<"$asdf_dir/version.txt")"
+  if [ -d "$asdf_dir/.git" ]; then
+    git_rev="$(git --git-dir "$asdf_dir/.git" rev-parse --short HEAD)"
     printf "%s-%s\n" "$version" "$git_rev"
   else
     printf "%s\n" "$version"
@@ -32,21 +34,24 @@ asdf_data_dir() {
   elif [ -n "$HOME" ]; then
     data_dir="$HOME/.asdf"
   else
-    data_dir=$(asdf_dir)
+    get_asdf_dir
+    data_dir=$REPLY
   fi
 
   printf "%s\n" "$data_dir"
 }
 
-asdf_dir() {
+get_asdf_dir() {
+  unset -v REPLY
+
   if [ -z "$ASDF_DIR" ]; then
     local current_script_path=${BASH_SOURCE[0]}
-    printf '%s\n' "$(
+    REPLY=$(
       cd -- "$(dirname "$(dirname "$current_script_path")")" || exit
       printf '%s\n' "$PWD"
-    )"
+    )
   else
-    printf '%s\n' "$ASDF_DIR"
+    REPLY="$ASDF_DIR"
   fi
 }
 
@@ -391,7 +396,9 @@ get_asdf_config_value() {
   local key=$1
   local config_path=
   config_path=$(asdf_config_file)
-  local default_config_path=${ASDF_CONFIG_DEFAULT_FILE:-"$(asdf_dir)/defaults"}
+  get_asdf_dir
+  local asdf_dir=$REPLY
+  local default_config_path=${ASDF_CONFIG_DEFAULT_FILE:-"$asdf_dir/defaults"}
 
   local local_config_path
   local_config_path="$(find_file_upwards ".asdfrc")"
@@ -701,7 +708,9 @@ get_shim_versions() {
 
 preset_versions() {
   shim_name=$1
-  shim_plugin_versions "${shim_name}" | cut -d' ' -f 1 | uniq | xargs -IPLUGIN bash -c ". $(asdf_dir)/lib/utils.bash; printf \"%s %s\n\" PLUGIN \$(get_preset_version_for PLUGIN)"
+  get_asdf_dir
+  local asdf_dir=$REPLY
+  shim_plugin_versions "${shim_name}" | cut -d' ' -f 1 | uniq | xargs -IPLUGIN bash -c ". $asdf_dir/lib/utils.bash; printf \"%s %s\n\" PLUGIN \$(get_preset_version_for PLUGIN)"
 }
 
 select_from_preset_version() {
diff --git a/test/test_helpers.bash b/test/test_helpers.bash
index 8a23931..af3d7a7 100644
--- a/test/test_helpers.bash
+++ b/test/test_helpers.bash
@@ -109,5 +109,8 @@ clean_asdf_dir() {
 
 setup_repo() {
   cp -r "$BATS_TEST_DIRNAME/fixtures/dummy_plugins_repo" "$ASDF_DIR/repository"
-  touch "$(asdf_dir)/tmp/repo-updated"
+  get_asdf_dir
+  local asdf_dir=$REPLY
+  echo j "$asdf_dir" >&2
+  touch "$asdf_dir/tmp/repo-updated"
 }
